### DISCUSSION TITLE
Different Solutions Running in Serial v. Parallel Using Python Wrapper w/ Mesh Deformation

### DISCUSSION THREAD

**Discussion by j-signorelli** [su2code/SU2#1931] [Q&A] [answered]
Hi,

I am obtaining quite different solutions in parallel v. in serial when using the Python wrapper with mesh deformation. I set up a simple test case to aid in helping resolve this issue. I have attached my Python script as a .txt file at the bottom.

This simple test case is just flow in a channel, where I use the Python wrapper to enforce an oscillating wall. I use an initial flowfield from a steady, nondeforming solution. Below is the video of the pressure through time:

https://user-images.githubusercontent.com/116454159/219805509-6c74e074-943e-410e-b497-e81420ea84b7.mp4

When I run the same setup but on 3 processes, I get the following result:

https://user-images.githubusercontent.com/116454159/219806456-ac445cb7-b761-4327-aca7-ebda193bd10f.mp4

There are very unusually sharp pressure gradients at the side as can be seen. This problem is occurring directly at the interface of domains decomposed onto different ranks:

![rank](https://user-images.githubusercontent.com/116454159/219807460-aec510d8-c1ad-4e24-9a62-2758ae060580.png)

Is there something I am missing in the Python file? Any help is **really** appreciated!

[SU2_FSI_TestCase_PythonFile.txt](https://github.com/su2code/SU2/files/10772070/SU2_FSI_TestCase_PythonFile.txt)

**Comment by pcarruscag** [comment]
What branch are you using? How do you compile it? Run it (python script)?

**Reply by j-signorelli** [reply]
Hi @pcarruscag , Thanks a ton for the quick response. I am using master branch v7.5.1.  I compile with:

```
export CC=$(which mpicc)
export CXX=$(which mpicxx)
export MPICC=$(which mpicc)
export MPICXX=$(which mpicxx)

./meson.py build -Dcustom-mpi=true  -Denable-pywrapper=true -Denable-mkl=true -Dmkl_root=$MKL_DIR --prefix=~/software/SU2/install/
./ninja -C build install

```

I run by:
```
mpirun -n 3 python3 SU2_FSI_TestCase.py -f configfile.cfg --parallel
```

**Reply by j-signorelli** [reply]
I should note I am using Intel MPI. I haven't been able to get SU2 to work with it without setting up "custom MPI" like this.

**Reply by pcarruscag** [reply]
I'm not sure this is right
```
    # Obtain indices of all vertices that are being worked on on this rank
    for iVertex in range(nVertex_MovingMarker):
        if not SU2Driver.IsAHaloNode(MovingMarkerID, iVertex):
            iVertices_MovingMarker_PHYS.append(int(iVertex))
```
You need to set the displacement for all vertices (including Halos) or call a function of the wrapper to communicate them (CommunicateMeshDisplacements ? , it should be easy to find if you run dir(SU2Driver). But in your case just loop over all vertices.
Let me know if that works.

**Reply by j-signorelli** [reply]
Thanks @pcarruscag but sadly I have already tried setting every vertex including halos without any luck, and just tried again to confirm once more.

I also do call `SU2Driver.CommunicateMeshDisplacement()` for good measure in the Python file I attached, but this made no difference as mesh displacements are already communicated in the beginning of `CMeshSolver::DeformMesh`, which is called in `CSinglezoneDriver::DynamicMeshUpdate`, which is called in `CSinglezoneDriver::Preprocess`. 

Please let me know if you have any other suggestions! I am not sure if this may be a bug in SU2.

**Comment by pcarruscag** [comment]
I see, the deformation looks smooth in both videos.
Is the flow solver converging to the same residuals in serial vs parallel?
Can you upload the screen output in both cases, the mesh and config would also help.

**Reply by j-signorelli** [reply]
Of course, thanks so much @pcarruscag ! I can confirm that the deformation in the output is smooth:

![image](https://user-images.githubusercontent.com/116454159/219874505-4866e63e-23c1-45c3-b12d-8852ec219a31.png)

I also did verify that the flow solver is converging to the same residuals in serial v. parallel. My convergence criteria is an RMS_ENERGY < 1e-1.5. Attached is screen output for 3 procs, 1 proc, the config file, the mesh, and the initial flow dat file, all as text files so I can upload them here.

[3cores.txt](https://github.com/su2code/SU2/files/10774639/3cores.txt)
[1core.txt](https://github.com/su2code/SU2/files/10774643/1core.txt)
[config_file_testcase.txt](https://github.com/su2code/SU2/files/10774644/config_file_testcase.txt)
[fluidMesh.txt](https://github.com/su2code/SU2/files/10774645/fluidMesh.txt)
[initial_flow_00000.txt](https://github.com/su2code/SU2/files/10774646/initial_flow_00000.txt)

**Reply by pcarruscag** [reply]
I'll have a look but if the mach number of 0.01 matches the velocities you are getting that may be the root of the problem. Compressible solvers without low speed preconditioning are known for not very good solutions.
Have you tried the incompressible solver?

**Reply by j-signorelli** [reply]
Just tried it, same problem sadly:

1 process:
![1proc_p](https://user-images.githubusercontent.com/116454159/219892881-da114c0e-d778-4de9-afd7-24bee721a484.png)


3 processes:
![3proc_p](https://user-images.githubusercontent.com/116454159/219892902-cb1df9b5-1049-4597-8ef7-7a547f979268.png)

![3proc_rank](https://user-images.githubusercontent.com/116454159/219892921-f7601193-24b0-4df8-b54f-fe5256357928.png)


Thanks so much for looking into this @pcarruscag

**Comment by pcarruscag** [comment, marked as answer]
Took a bit but I found it https://github.com/su2code/SU2/pull/1933
It was just a problem with euler markers + deforming markers, not related to the wrapper or MPI.
We recently changed the python wrapper for version 8... so your script will not work directly with the branch in the PR.
You can use this script [run.py.txt](https://github.com/su2code/SU2/files/10775614/run.py.txt) or git-cherrypick my commit into master and continue using yours.

Can we use your example as a regression test?

**Reply by j-signorelli** [reply]
Thank you so much for figuring this out! You can absolutely use the example.
